extensions/amp-story/1.0/amp-story-tooltip.js

```diff
@@ -0,0 +1,401 @@
+    const existingTooltipIcon = childElement(
+        dev().assertElement(this.tooltip_),
+        el => el.classList.contains('i-amphtml-story-tooltip-icon'));
+
+    if (!iconSrc) {
+      if (existingTooltipIcon) {
+        removeElement(existingTooltipIcon);
+      }
+      return;
+    }
+
+    if (existingTooltipIcon && existingTooltipIcon.firstElementChild) {
+      addAttributesToElement(existingTooltipIcon.firstElementChild,
+          dict({'src': iconSrc}));
+    } else {
+      const html = htmlFor(this.win_.document);
+      const iconEl =
+        html`
+        <div class="i-amphtml-story-tooltip-icon">
+          <img ref="icon">
+          </img>
+        </div>`;
+      const {icon} = htmlRefs(iconEl);
+      addAttributesToElement(icon, dict({'src': iconSrc}));
+      this.tooltip_.insertBefore(iconEl, this.tooltip_.firstChild);
+    }
+  }
+
+  /**
+   * Positions tooltip and its pointing arrow according to the position of the
+   * clicked element.
+   * @param {!Element} clickedEl
+   * @private
+   */
+  positionTooltip_(clickedEl) {
+    let clickedRect, storyWidth, storyHeight, tooltipWidth, tooltipTop,
+        tooltipLeft, arrowLeftOffset, arrowOnTop = false;
+
+    this.resources_.measureMutateElement(this.storyEl_,
+        /** measure */
+        () => {
+          clickedRect = clickedEl./*OK*/getBoundingClientRect();
+          storyHeight = this.storyEl_./*OK*/offsetHeight;
+
+          // Vertical positioning.
+          if (clickedRect.top > MIN_VERTICAL_SPACE) { // Tooltip fits above clicked element.
+            tooltipTop = clickedRect.top - MIN_VERTICAL_SPACE;
+          } else if (storyHeight - clickedRect.bottom > MIN_VERTICAL_SPACE) { // Tooltip fits below clicked element. Place arrow on top of the tooltip.
+            tooltipTop = clickedRect.bottom + EDGE_PADDING;
+            arrowOnTop = true;
+          } else { // Element takes whole vertical space. Place tooltip on the middle.
+            tooltipTop = storyHeight / 2;
+          }
+
+          tooltipWidth = this.tooltip_./*OK*/offsetWidth;
+          storyWidth = this.storyEl_./*OK*/offsetWidth;
+
+          // Horizontal positioning.
+          const elCenter = (clickedRect.width / 2) + clickedRect.left;
+          tooltipLeft = elCenter - (tooltipWidth / 2);
+          const maxHorizontal = (storyWidth - tooltipWidth - EDGE_PADDING);
+          // Make sure tooltip is not out of screen.
+          tooltipLeft = Math.min(tooltipLeft, maxHorizontal);
+          tooltipLeft = Math.max(EDGE_PADDING, tooltipLeft);
+          // Position tooltip arrow horizontally depending on clicked
+          // element's center in relation to the screen width.
+          arrowLeftOffset = elCenter / storyWidth * 100;
+          arrowLeftOffset = Math.min(arrowLeftOffset, 80);
+          arrowLeftOffset = Math.max(arrowLeftOffset, 0);
+        },
+        /** mutate */
+        () => {
+          // Arrow on top or bottom of tooltip.
+          arrowOnTop ?
```

Comment:
If the arrow is on top, we should also change the `translate3d` to slide down, instead of sliding up :))